File: R/dnam_network_clock.R
# DNAm Network Clock, age prediction from methylation beta values (R)
#
# This is a single, self contained file that uses only base R. There is
# nothing to install. Source it into your own script:
#
#     source("path/to/dnam_network_clock.R")
#     ages <- predict_age(betas)
#
# where `betas` is a data.frame or matrix with samples as ROWS and CpG IDs
# (Illumina cg identifiers) as COLUMN names. Missing CpGs are tolerated.
#
# The clock is a K=60 principal component ridge regression on per module
# medians of methylation beta values. See README.md for the full methodology,
# the trained model metrics, and the paper reference.
#
# Author: Anton Carcedo
# Licence: MIT

# Record the directory this file lives in, at source time, so the weights
# folder can be found automatically later. Scans the call frames for the path
# that source() records, with a fallback for Rscript.
.dnam_this_script_dir <- function() {
  for (i in seq_len(sys.nframe())) {
    of <- sys.frame(i)$ofile
    if (!is.null(of)) return(normalizePath(dirname(of)))
  }
  args <- commandArgs(trailingOnly = FALSE)
  fa <- grep("^--file=", args, value = TRUE)
  if (length(fa) > 0) {
    return(normalizePath(dirname(sub("^--file=", "", fa[1]))))
  }
  NA_character_
}
.DNAM_SCRIPT_DIR <- .dnam_this_script_dir()


# Locate the "weights" folder.
#
# Tries, in order: getOption("dnam_clock.weights_dir"), a "weights" folder next
# to this file, a "weights" folder one level up (the repository layout, where
# this file sits in R/ and the weights sit in ../weights), and "weights" in the
# current working directory.
.dnam_locate_weights_dir <- function() {
  opt <- getOption("dnam_clock.weights_dir", NULL)
  if (!is.null(opt) && dir.exists(opt)) return(normalizePath(opt))

  candidates <- character(0)
  if (!is.na(.DNAM_SCRIPT_DIR)) {
    candidates <- c(candidates,
                    file.path(.DNAM_SCRIPT_DIR, "weights"),
                    file.path(dirname(.DNAM_SCRIPT_DIR), "weights"))
  }
  candidates <- c(candidates, file.path(getwd(), "weights"))

  for (p in candidates) if (dir.exists(p)) return(normalizePath(p))

  stop("DNAm Network Clock: could not locate the 'weights' directory. ",
       "Set options(dnam_clock.weights_dir = '/path/to/weights') ",
       "or pass weights_dir= to predict_age().")
}

# Lightweight string hash for cache keys (avoids a 'digest' package dependency).
.dnam_digest_str <- function(s) {
  chars <- strsplit(s, "")[[1]]
  paste0("h", as.character(sum(utf8ToInt(s) * seq_along(chars))))
}

# Load and cache the K=60 clock weights from the CSV files.
.dnam_load_weights <- function(weights_dir = NULL) {
  if (is.null(weights_dir)) weights_dir <- .dnam_locate_weights_dir()
  cache_key <- paste0(".dnam_weights_cache_", .dnam_digest_str(weights_dir))
  if (exists(cache_key, envir = .GlobalEnv)) {
    return(get(cache_key, envir = .GlobalEnv))
  }

  defs   <- read.csv(file.path(weights_dir, "module_definitions.csv"),
                     stringsAsFactors = FALSE)
  params <- read.csv(file.path(weights_dir, "module_params.csv"),
                     stringsAsFactors = FALSE)
  comps  <- read.csv(file.path(weights_dir, "pca_components_k60.csv"),
                     stringsAsFactors = FALSE, check.names = FALSE)
  ridge  <- read.csv(file.path(weights_dir, "ridge_k60.csv"),
                     stringsAsFactors = FALSE)

  module_order <- params$Module                   # canonical column order

  # Split CpGs by module, then reorder to module_order.
  defs_split <- split(defs$CpG, defs$Module)
  module_to_cpgs <- defs_split[as.character(module_order)]

  # Strategy, one per module (take the first row).
  strat_per_module <- tapply(defs$Strategy, defs$Module, function(x) x[1])
  module_to_strategy <- as.character(strat_per_module[as.character(module_order)])

  # PCA components: 60 rows x 4595 modules, drop the 'PC' label column.
  pca_components <- as.matrix(comps[, -1, drop = FALSE])

  # Ridge coefficients plus intercept.
  pc_mask    <- grepl("^PC", ridge$name)
  pc_coefs   <- ridge$value[pc_mask]
  intercept  <- ridge$value[ridge$name == "intercept"]

  stopifnot(length(module_order)     == 4595,
            length(module_to_cpgs)   == 4595,
            length(pc_coefs)         == 60,
            nrow(pca_components)      == 60,
            ncol(pca_components)      == 4595,
            length(intercept)         == 1)

  weights <- list(
    module_order       = module_order,
    module_to_cpgs     = module_to_cpgs,
    module_to_strategy = module_to_strategy,
    scaler_mean        = params$scaler_mean,
    scaler_scale       = params$scaler_scale,
    pca_mean           = params$pca_mean,
    pca_components     = pca_components,
    pc_coefs           = pc_coefs,
    intercept          = intercept,
    n_modules          = length(module_order),
    K                  = length(pc_coefs)
  )
  assign(cache_key, weights, envir = .GlobalEnv)
  weights
}


# Build the samples x modules matrix from samples x CpGs.
# Missing CpGs and NA values are tolerated.
.dnam_build_module_matrix <- function(betas, weights, warn_missing_above = 0.05) {
  if (is.data.frame(betas)) betas <- as.matrix(betas)
  if (!is.matrix(betas))
    stop("betas must be a matrix or data.frame (samples x CpGs).")
  if (is.null(colnames(betas)))
    stop("betas must have CpG IDs as column names.")

  n_samples <- nrow(betas)
  n_modules <- weights$n_modules
  X <- matrix(0.0, nrow = n_samples, ncol = n_modules)
  available <- colnames(betas)

  n_missing_modules     <- 0L
  n_singleton_fallback  <- 0L
  n_partial_median      <- 0L
  n_cpgs_used           <- 0L
  n_cpgs_needed         <- 0L

  for (j in seq_len(n_modules)) {
    cpgs     <- weights$module_to_cpgs[[j]]
    strategy <- weights$module_to_strategy[j]
    n_cpgs_needed <- n_cpgs_needed + length(cpgs)
    present  <- intersect(cpgs, available)
    n_cpgs_used <- n_cpgs_used + length(present)

    if (strategy == "singleton") {
      if (length(present) == 1L) {
        X[, j] <- betas[, present]
      } else {
        X[, j] <- weights$scaler_mean[j]
        n_singleton_fallback <- n_singleton_fallback + 1L
      }
    } else {  # 'median'
      if (length(present) == 0L) {
        X[, j] <- weights$scaler_mean[j]
        n_missing_modules <- n_missing_modules + 1L
      } else {
        if (length(present) < length(cpgs)) n_partial_median <- n_partial_median + 1L
        if (length(present) == 1L) {
          X[, j] <- betas[, present]
        } else {
          X[, j] <- apply(betas[, present, drop = FALSE], 1, median, na.rm = TRUE)
        }
        nan_rows <- is.na(X[, j])
        if (any(nan_rows)) X[nan_rows, j] <- weights$scaler_mean[j]
      }
    }
  }

  coverage <- if (n_cpgs_needed > 0) n_cpgs_used / n_cpgs_needed else 0
  missing_frac <- 1 - coverage
  if (missing_frac > warn_missing_above) {
    warning(sprintf(
      paste("DNAm Network Clock: %.1f%% of the clock's CpGs are missing from",
            "the input (%d/%d present). %d modules were filled with the",
            "training mean. Predictions may be biased toward the training",
            "set mean age."),
      missing_frac * 100, n_cpgs_used, n_cpgs_needed,
      n_missing_modules + n_singleton_fallback
    ), call. = FALSE)
  }

  list(
    X = X,
    diagnostics = list(
      n_modules = n_modules,
      n_cpgs_used = n_cpgs_used,
      n_cpgs_needed = n_cpgs_needed,
      cpg_coverage = coverage,
      n_modules_filled_with_training_mean = n_missing_modules + n_singleton_fallback,
      n_partial_median_modules = n_partial_median
    )
  )
}


# Predict chronological age from a methylation beta matrix.
#
# betas              A matrix or data.frame with samples as rows and CpG IDs
#                    (Illumina cg identifiers) as column names.
# weights_dir        Optional path to a directory of weight CSVs. If NULL, the
#                    "weights" folder is located automatically.
# return_diagnostics If TRUE, return a list(ages, diagnostics).
#
# Returns a numeric vector of predicted ages (years), one per sample in the row
# order of `betas`. If return_diagnostics=TRUE, a list with elements `ages` and
# `diagnostics`.
predict_age <- function(betas, weights_dir = NULL, return_diagnostics = FALSE) {
  if (is.null(rownames(betas)) && is.data.frame(betas))
    rownames(betas) <- seq_len(nrow(betas))

  weights <- .dnam_load_weights(weights_dir)

  built <- .dnam_build_module_matrix(betas, weights)
  X <- built$X

  # standardise, then centre, then project, then ridge.
  X <- sweep(X, 2, weights$scaler_mean,  "-")
  X <- sweep(X, 2, weights$scaler_scale, "/")
  X <- sweep(X, 2, weights$pca_mean,     "-")
  PC_scores <- X %*% t(weights$pca_components)               # n x 60
  ages <- as.numeric(PC_scores %*% weights$pc_coefs + weights$intercept)

  if (return_diagnostics) list(ages = ages, diagnostics = built$diagnostics) else ages
}

# Return the trained clock's metadata.
clock_info <- function(weights_dir = NULL) {
  if (is.null(weights_dir)) weights_dir <- .dnam_locate_weights_dir()
  meta_path <- file.path(weights_dir, "metadata.json")
  if (!file.exists(meta_path))
    stop("metadata.json not found in ", weights_dir)
  raw <- paste(readLines(meta_path, warn = FALSE), collapse = "\n")
  if (requireNamespace("jsonlite", quietly = TRUE)) {
    jsonlite::fromJSON(raw)
  } else {
    message("Install 'jsonlite' for a structured metadata object, ",
            "returning the raw JSON string.")
    raw
  }
}

File: python/dnam_network_clock.py
"""DNAm Network Clock, age prediction from methylation beta values (Python).

This is a single, self contained file. There is nothing to install. Either
copy it into your own project, or keep it next to the "weights" folder and
import it:

    from dnam_network_clock import predict_age
    ages = predict_age(betas)        # betas: pandas DataFrame, samples x CpGs

The clock is a K=60 principal component ridge regression on per module medians
of methylation beta values. Modules were discovered by Infomap clustering of a
CpG to CpG Spearman correlation network (threshold tau=0.70) over 7,328 CpGs
pre filtered by absolute Spearman rho with age >= 0.35, on a training set of
1,917 whole blood methylation samples spanning ages 18 to 94 across 12 studies.
See README.md for the full methodology, the trained model metrics, and the
paper reference.

Requires: numpy >= 1.20, pandas >= 1.3 (Python >= 3.9).
Licence: MIT.
"""
from __future__ import annotations

import json
import warnings
from pathlib import Path
from typing import Optional, Union

import numpy as np
import pandas as pd


# Weight loading, cached once per process.
_WEIGHTS_CACHE: dict = {}


def _locate_weights_dir() -> Path:
    """Find the "weights" folder shipped alongside this file.

    Searched in order: the current working directory, the folder this file
    lives in, and the parent of that folder (the repository layout, where the
    code sits in python/ and the weights sit in ../weights). If none is found,
    pass weights_dir= explicitly to predict_age().
    """
    here = Path(__file__).resolve().parent
    candidates = [
        Path.cwd() / "weights",
        here / "weights",
        here.parent / "weights",
    ]
    for c in candidates:
        if c.is_dir():
            return c
    raise FileNotFoundError(
        "Could not locate the 'weights' directory. Pass weights_dir= "
        "explicitly to predict_age(), e.g. predict_age(betas, weights_dir='/path/to/weights')."
    )


def _load_weights(weights_dir: Optional[Union[str, Path]] = None) -> dict:
    """Load and cache the K=60 clock weights from the CSV files in weights_dir."""
    key = str(weights_dir) if weights_dir is not None else "<auto>"
    if key in _WEIGHTS_CACHE:
        return _WEIGHTS_CACHE[key]

    d = Path(weights_dir) if weights_dir is not None else _locate_weights_dir()

    defs   = pd.read_csv(d / "module_definitions.csv")
    params = pd.read_csv(d / "module_params.csv")
    comps  = pd.read_csv(d / "pca_components_k60.csv")
    ridge  = pd.read_csv(d / "ridge_k60.csv")
    with open(d / "metadata.json") as f:
        meta = json.load(f)

    # Module column order is given by module_params.csv.
    module_order = params["Module"].to_numpy()
    assert len(module_order) == meta["n_modules"]

    # Per module lookup: Module -> (list of CpGs, strategy).
    grouped = defs.groupby("Module", sort=False)
    module_to_cpgs = {int(m): g["CpG"].tolist() for m, g in grouped}
    module_to_strategy = {int(m): g["Strategy"].iloc[0] for m, g in grouped}

    # PCA components: columns are module IDs in module_order, rows are PCs.
    pca_components = comps.drop(columns=["PC"]).to_numpy()
    assert pca_components.shape[0] == meta["K"]
    assert pca_components.shape[1] == meta["n_modules"]

    # Ridge: PC1..PCK rows, plus an intercept row.
    pc_rows = ridge[ridge["name"].str.startswith("PC")]
    pc_coefs = pc_rows["value"].to_numpy()
    intercept = float(ridge.loc[ridge["name"] == "intercept", "value"].iloc[0])
    assert pc_coefs.shape == (meta["K"],)

    weights = {
        "module_order":       module_order,
        "module_to_cpgs":     module_to_cpgs,
        "module_to_strategy": module_to_strategy,
        "scaler_mean":        params["scaler_mean"].to_numpy(),
        "scaler_scale":       params["scaler_scale"].to_numpy(),
        "pca_mean":           params["pca_mean"].to_numpy(),
        "pca_components":     pca_components,
        "pc_coefs":           pc_coefs,
        "intercept":          intercept,
        "metadata":           meta,
    }
    _WEIGHTS_CACHE[key] = weights
    return weights


# Per module aggregation: build samples x modules matrix from samples x CpGs.
def _build_module_matrix(betas: pd.DataFrame, weights: dict,
                         warn_missing_above: float = 0.05) -> "tuple[np.ndarray, dict]":
    """Aggregate the input beta matrix (samples x CpGs) into a samples x modules
    matrix using each module's stored strategy.

    Missing CpGs are tolerated. 'median' modules use the median of the
    available CpGs per sample, falling back to the training scaler_mean when a
    module has no input CpGs at all. 'singleton' modules use the single CpG's
    value, or scaler_mean if that CpG is missing.
    """
    n_samples = betas.shape[0]
    module_order = weights["module_order"]
    n_modules = len(module_order)
    scaler_mean = weights["scaler_mean"]

    X = np.empty((n_samples, n_modules), dtype=np.float64)
    available_cpgs = set(betas.columns.astype(str))

    n_missing_modules = 0
    n_singleton_fallback = 0
    n_partial_median = 0
    n_cpgs_used = 0
    n_cpgs_needed = 0

    for j, mid in enumerate(module_order):
        cpgs = weights["module_to_cpgs"][int(mid)]
        strategy = weights["module_to_strategy"][int(mid)]
        n_cpgs_needed += len(cpgs)
        present = [c for c in cpgs if c in available_cpgs]
        n_cpgs_used += len(present)

        if strategy == "singleton":
            if present:
                X[:, j] = betas[present[0]].to_numpy()
            else:
                X[:, j] = scaler_mean[j]
                n_singleton_fallback += 1
        else:  # 'median'
            if not present:
                X[:, j] = scaler_mean[j]
                n_missing_modules += 1
            else:
                if len(present) < len(cpgs):
                    n_partial_median += 1
                # Median across the available CpGs for each sample.
                X[:, j] = np.nanmedian(betas[present].to_numpy(), axis=1)
                # If every CpG of a module was NaN for a sample, fall back.
                nan_mask = np.isnan(X[:, j])
                if nan_mask.any():
                    X[nan_mask, j] = scaler_mean[j]

    coverage = n_cpgs_used / n_cpgs_needed if n_cpgs_needed else 0.0
    diagnostics = {
        "n_modules": n_modules,
        "n_cpgs_used": n_cpgs_used,
        "n_cpgs_needed": n_cpgs_needed,
        "cpg_coverage": coverage,
        "n_modules_filled_with_training_mean": n_missing_modules + n_singleton_fallback,
        "n_partial_median_modules": n_partial_median,
    }
    missing_frac = 1.0 - coverage
    if missing_frac > warn_missing_above:
        warnings.warn(
            f"DNAm Network Clock: {missing_frac*100:.1f}% of the clock's CpGs are "
            f"missing from the input ({n_cpgs_used:,}/{n_cpgs_needed:,} present). "
            f"{diagnostics['n_modules_filled_with_training_mean']:,} modules were "
            f"filled with the training mean. Predictions may be biased toward "
            f"the training set mean age.",
            UserWarning, stacklevel=2,
        )
    return X, diagnostics


# Public API.
def predict_age(
    betas: pd.DataFrame,
    weights_dir: Optional[Union[str, Path]] = None,
    return_diagnostics: bool = False,
) -> Union[np.ndarray, "tuple[np.ndarray, dict]"]:
    """Predict chronological age from a methylation beta value matrix.

    Parameters
    ----------
    betas
        Methylation beta values as a pandas DataFrame with samples as rows and
        CpG IDs (Illumina cg identifiers) as columns. Missing CpGs and within
        column NaN are tolerated, see Notes.
    weights_dir
        Optional path to a directory of weight CSVs. By default the "weights"
        folder shipped with this file is located automatically.
    return_diagnostics
        If True, also return a dict summarising CpG coverage and which modules
        were imputed from the training mean.

    Returns
    -------
    ndarray of shape (n_samples,)
        Predicted ages in years, in the same order as betas.index. If
        return_diagnostics=True, a 2 tuple (ages, diagnostics).

    Notes
    -----
    The clock has 4,595 modules covering 7,328 CpGs. Each module is aggregated
    to a single value per sample by taking the median of the available CpGs
    (multi CpG modules) or using the single CpG (singleton modules). A module
    with no input CpGs at all is filled with the training time mean for that
    module. A warning is emitted if more than 5% of the clock's CpGs are
    missing. The model is trained on whole blood methylation, performance on
    other tissues has not been characterised.

    Examples
    --------
    >>> import pandas as pd
    >>> from dnam_network_clock import predict_age
    >>> betas = pd.read_csv("my_methylation.csv", index_col=0)   # samples x CpGs
    >>> ages = predict_age(betas)
    >>> ages[:5]
    array([42.1, 67.8, 28.3, 55.4, 71.9])
    """
    if not isinstance(betas, pd.DataFrame):
        raise TypeError(
            "betas must be a pandas DataFrame with samples as rows and CpG "
            "IDs as columns. Got: " + type(betas).__name__
        )
    if betas.shape[0] == 0:
        raise ValueError("betas has 0 rows (samples).")

    weights = _load_weights(weights_dir)

    # 1. samples x modules matrix
    X, diagnostics = _build_module_matrix(betas, weights)

    # 2. standardise
    X = (X - weights["scaler_mean"]) / weights["scaler_scale"]

    # 3. centre for PCA, then project onto the top K PCs
    X = X - weights["pca_mean"]
    PC_scores = X @ weights["pca_components"].T          # (n_samples, 60)

    # 4. ridge
    ages = PC_scores @ weights["pc_coefs"] + weights["intercept"]

    if return_diagnostics:
        return ages, diagnostics
    return ages


def clock_info(weights_dir: Optional[Union[str, Path]] = None) -> dict:
    """Return the trained clock's metadata (K, module counts, training metrics)."""
    return dict(_load_weights(weights_dir)["metadata"])


__all__ = ["predict_age", "clock_info"]
